resolver/resolver.go

```diff
@@ -3507,33 +3494,85 @@ func (r *Resolver) createCELEnv(msg *Message) (*cel.Env, error) {
 	return env, nil
 }
 
-func (r *Resolver) getEnvMessage(msg *Message) *Message {
+func (r *Resolver) buildEnvMessage(msg *Message, svcMsgSet map[*Service]map[*Message]struct{}) (*Message, error) {
 	if msg.File == nil {
-		return nil
+		return nil, nil
 	}
 	if msg.File.Package == nil {
-		return nil
+		return nil, nil
 	}
-	var envSvcs []*Service
-	for _, file := range msg.File.Package.Files {
-		for _, svc := range file.Services {
-			if svc.Rule == nil {
-				continue
-			}
-			if svc.Rule.Env == nil {
-				continue
+
+	// Examine all the services where the message is used and collect all the essential information
+	// to calculate the intersection set of each service's environment variables.
+	svcEnvVarMap := map[*Service]map[string]*EnvVar{}
+	envVarNameSet := map[string]struct{}{}
+	for svc, msgSet := range svcMsgSet {
+		if _, exist := msgSet[msg]; exist {
```

Comment:
nits: Since `exists` is used in all other places, please align this one as well if possible.